proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,354 @@
+---
+simd: '0016'
+title: Application Fees (Write-lock fees)
+authors:
+  - Godmode Galactus (Mango Markets)
+  - Maximilian Schneider (Mango Markets)
+category: Standard
+type: Core, Fees
+status: Draft
+created: 2022-12-23
+feature:
+---
+
+
+## Summary
+
+This SIMD will discuss additional fees called Application Fees or write lock fees.
+These fees are decided and set by the dapp developer to interact with the dapp usually by
+write-locking one of the accounts. Dapp developers then can decide to rebate these fees
+if a user interacts with the dapp as intended and disincentivize the users which do not
+interact with the app as intentended. These fees will be applied even if the transaction
+eventually fails. These fees will be collected on the same writable account and the
+authority can do lamport transfers to recover these fees.
```

Comment:
Yes authority is the owner of the account.

Reply:
Sorry, it's still not totally clear in my mind. Here's a scenario that I made up that may help clarify:

Suppose I write a faucet program with an instruction `Airdrop` that takes the user's token account and the program's faucet token account. If you pass the right value in the instruction data, it transfers some tokens from the faucet account to your account.
If I want to use the system you're proposing to protect the faucet token account, what accounts get passed to `Rebate`? 

The owner of the account is the token program. The "authority" from the Token program's perspective is some PDA of my faucet program, but the runtime has no understanding of that kind of authority. Depending on how I created the token account, it might be a PDA or it might be a throwaway account.

I think the easiest way is to say the account from which the rebate is coming must be the signer for the `Rebate` instruction. In the scenario above, that means that it may not be possible to protect the token account (if it isn't a PDA), but I think that's potentially okay. Accounts larger than 10kb are never PDAs though. Is that an issue?

As far as I know, the only notion of having the `program owner` of the account sign is via an extra PDA. For example, the Rebate program could allow you to rebate any account owned by program X if `PDA(X, "RebateProgram")` is a signer on the instruction. That would allow you to rebate non-PDAs that are owned by the program. It doesn't enable the faucet scenario, but maybe that's okay.

Reply:
In this case the faucet account will have application fees, so normally for pda's they are itself their authority (owner), so we can just call invoke_signed using signer seeds.

Check this code where the check for rebate is added:
https://github.com/blockworks-foundation/solana/blob/8782eb2c3756f2714e3241c36f02d717f83baf5e/programs/application-fees/src/processor.rs#L99

Check this code in an example smart contract:
https://github.com/godmodegalactus/paper-clip-maximizer/blob/6cfb4387033e37eccdfb6c5ef7ed6fc0331bddad/programs/paper-clip-maximizer/src/lib.rs#L119

I changed from owner to authority after an suggession here:
https://github.com/solana-foundation/solana-improvement-documents/pull/16#discussion_r1101079462

Reply:
Hey,

Do you agree to close this point or still some confusion? 
After my reply here 
https://github.com/solana-foundation/solana-improvement-documents/pull/16#issuecomment-1435531441